contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,196 @@
+# DaemonSet Updates
+
+**Author**: @madhusudancs, @lukaszo, @janetkuo
+
+**Status**: Proposal
+
+## Abstract
+
+A proposal for adding the update feature to `DaemonSet`.
+
+Users already can update a `DaemonSet`, which will not cause changes to its
+subsequent pods, until those pods are killed. In this proposal, we plan to add
+a "RollingUpdate" strategy which allows DaemonSet to downstream its changes to
+pods. 
+
+## Implementation 
+
+### API Object 
+
+To enable DaemonSet upgrades, the `DaemonSet` API object will have the following
+structure:
+
+```go 
+type DaemonSetUpdateStrategy struct {
+	// Type of daemon set update. Can be "RollingUpdate" or "OnDelete". 
+	// Default is OnDelete.
+	Type DaemonSetUpdateStrategyType
+
+	// Rolling update config params. Present only if DaemonSetUpdateStrategy =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be. Same as DeploymentStrategy.RollingUpdate.
+	RollingUpdate *RollingUpdateDaemonSet
+}
+
+type DaemonSetUpdateStrategyType string
+
+const (
+	// Replace the old daemons by new ones using rolling update i.e replace them on each node one after the other.
+	RollingUpdateDaemonSetStrategyType DaemonSetUpdateStrategyType = "RollingUpdate"
+        
+	// Replace the old daemons only when it's killed
+	OnDeleteDaemonSetStrategyType DaemonSetUpdateStrategyType = "OnDelete"
+)
+
+// Spec to control the desired behavior of daemon set rolling update.
+type RollingUpdateDaemonSet struct {
+	// The maximum number of DaemonSet pods that can be unavailable during the
```

Comment:
@erictune is it always the case that we don't care about minimum available daemon pods? I haven't thought about it before and I think I agree with you but I am not sure of all the posibilities. Do services in front of daemon sets work?

> What if the purpose of the rollout is to fix them by updating the pod template?

The rollout process should be able to scale down broken pods no matter what.

Parallelism actually sounds like a better approach to control the speed of the rollout than MaxUnavailable but if we don't care about availability I fail to understand why we care about the speed of the rollout anyway. I think we may want to start with a canary mechanism for this (update one pod and wait for it to come up before proceeding to update all of the rest).